### ISSUE TITLE
Handling of class-decorators with overloaded `__call__` used with overloaded methods

### ISSUE THREAD

**Issue by Avasam** [microsoft/pyright#7167]
**Describe the bug**
Classes with `__call__` overloaded used to decorate methods also overloaded (`@overload` decorator) cause only the last overload to be kept

See code below for details. The current behaviour seems incorrect. I think this should either be supported to expand to all overloads (I assume could be costly or complex), or explicitly disallowed (easy short-term fix). 

**Code or Screenshots**

```python
from collections.abc import Callable
from typing import Any, Generic, TypeVar, overload
from typing_extensions import ParamSpec, Self, reveal_type

_P = ParamSpec("_P")
_R = TypeVar("_R")

class DecoratorClass(Generic[_P, _R]):
    def __new__(cls, func: Callable[_P, _R] | None = None) -> Self: ...
    @overload
    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
    @overload
    def __call__(self, *args: Any, backend: str, **backend_kwargs: Any) -> _R: ...

@overload
@DecoratorClass
def overload_first(foo: str) -> None: ...
@overload
@DecoratorClass
def overload_first(foo: int) -> None: ...

@DecoratorClass
@overload
def overload_second(foo: str) -> None: ...
@DecoratorClass
@overload
def overload_second(foo: int) -> None: ...

_ = reveal_type(overload_first)
_ = reveal_type(overload_second)
```

pyright currently accepts the above and sees the type as such: (only the last method overload is kept, with the class overload)
```
Running Pyright (base configs) for Python 3.8...
9.6.2
Running: C:\Program Files\nodejs\npx.CMD pyright@1.1.342 stubs/networkx --pythonversion 3.8
e:\Users\Avasam\Documents\Git\typeshed\stubs\networkx\networkx\__init__.pyi
  e:\Users\Avasam\Documents\Git\typeshed\stubs\networkx\networkx\__init__.pyi:30:17 - information: Type of "overload_first" is "DecoratorClass[(foo: int), None]"
  e:\Users\Avasam\Documents\Git\typeshed\stubs\networkx\networkx\__init__.pyi:31:17 - information: Type of "overload_second" is "DecoratorClass[(foo: int), None]"
```

The Pylance language server shows the following overloads
![image](https://github.com/python/mypy/assets/1350584/7c52ff42-dfb4-43ff-8794-000afa2dfe7a)
![image](https://github.com/python/mypy/assets/1350584/220cb1c5-b3dd-42f8-95e9-4faa9afdf40a)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?
pyright: 1.1.342
pylance: v2024.1.102

**Additional information**
Mypy currently raises an error with an erroneous message: https://github.com/python/mypy/issues/16840

**Comment by Avasam** [comment]
As for the real world scenario where I encountered this: https://github.com/python/typeshed/pull/11336/files#diff-88b4b1d81d8a23e6e0f612acfb230bb8551893c0a56fed353784b2e744930430R48-R59

**Comment by erictraut** [comment]
I think your use of `overload` here is invalid, so I don't consider this a bug.

An `@overload` decorator should be applied to a function or method, but you're applying it to a non-function object of type `DecoratorClass`.

If you think this use case should be supported, the typing specification for `@overload` would need to be extended to define the rules for how it should be handled. You're certainly welcome to start a thread in the Python typing discussion forums to see if you could gather support for such an extension.

**Comment by Avasam** [comment]
Thanks for your answer.
I can definitely start a discussion to see if it's worth supporting this use case.
But even as something that is not currently supported, I find the current pyright behaviour to be odd. Could/should this have an associated diagnostic message rather than discarding the method overloads ?
